### PR TITLE
web: start web handler while TSDB is starting up

### PR DESCRIPTION
**Pull request by fabxc**
Just shifting around a few things so we can securely start the web handlers while TSDB is loading.
Some handlers no longer return 5xx before the server is marked as ready.

@Gouthamve

### CODE REVIEW

File: storage/tsdb/tsdb.go [hunk 1]
```diff
@@ -26,6 +29,55 @@ import (
 	tsdbLabels "github.com/prometheus/tsdb/labels"
 )
 
+var ErrNotReady = errors.New("TSDB not ready")
+
+type ReadyStorage struct {
+	mtx sync.RWMutex
```
Comment (fabxc) [inline storage/tsdb/tsdb.go]: The lock and `ErrNotReady` shouldn't really be necessary as no components should try to hit TSDB before it has started, but it's a bit safer and shouldn't cost too much.

File: storage/tsdb/tsdb.go [hunk 2]
```diff
@@ -66,8 +118,9 @@ func Open(path string, r prometheus.Registerer, opts *Options) (*tsdb.DB, error)
 			break
 		}
 	}
+	logger := log.NewLogfmtLogger(log.NewSyncWriter(os.Stdout))
```
Comment (fabxc) [inline storage/tsdb/tsdb.go]: Should be gone once we logging migration is done.

File: web/web.go [hunk 3]
```diff
@@ -205,34 +205,34 @@ func New(o *Options) *Handler {
...
+	router.Get("/heap", instrf("heap", dumpHeap))
 
 	router.Get("/metrics", prometheus.Handler().ServeHTTP)
 
 	router.Get("/federate", readyf(instrh("federate", httputil.CompressionHandler{
 		Handler: http.HandlerFunc(h.federation),
 	})))
 
-	router.Get("/consoles/*filepath", readyf(instrf("consoles", h.consoles)))
+	router.Get("/consoles/*filepath", instrf("consoles", h.consoles))
```
Comment (grobie) [inline web/web.go]: Why is it safe to allow any kind of console page while the server isn't ready yet?

Comment (fabxc) [reply web/web.go]: I thought it was using some JS to hit the query endpoint. If not, that should indeed be reverted.

Comment (brian-brazil) [reply web/web.go]: No, it has built-in query functionality. This should only be available once the TSDB is.

Comment (fabxc) [reply web/web.go]: Okay, reverted.

**Comment by fabxc** [comment]
Rebased against logging migration changes.

**Comment by fabxc** [comment]
@brian-brazil  @grobie any further concerns?

File: cmd/prometheus/main.go [hunk 4]
```diff
@@ -215,35 +215,17 @@ func main() {
...
+		fanoutStorage = storage.NewFanout(logger, localStorage, remoteStorage)
 	)
-	if err != nil {
-		level.Error(logger).Log("msg", "Opening TSDB failed", "err", err)
-		os.Exit(1)
-	}
-
-	level.Info(logger).Log("msg", "TSDB succesfully started")
-
-	remoteStorage := remote.NewStorage(log.With(logger, "component", "remote"))
```
Comment (grobie) [inline cmd/prometheus/main.go]: Where did the logger initialization go?

File: cmd/prometheus/main.go [hunk 5]
```diff
@@ -314,14 +298,29 @@ func main() {
 		}
 	}()
 
-	// Start all components. The order is NOT arbitrary.
-	defer func() {
-		if err := fanoutStorage.Close(); err != nil {
-			level.Error(logger).Log("msg", "Closing storage(s) failed", "err", err)
+	// Start all components while we wait for TSDB to open but only load
+	// initial config and mark ourselves as ready after it completed.
```
Comment (grobie) [inline cmd/prometheus/main.go]: Is it safe to start already the reload signal handler above?

File: cmd/prometheus/main.go (hunk 4 above)
Comment (fabxc) [reply cmd/prometheus/main.go]: Good catch, rebase issue.

File: cmd/prometheus/main.go (hunk 5 above)
Comment (fabxc) [reply cmd/prometheus/main.go]: You mean the goroutine reading HUP signals from the channel? That can be anywhere really. Just waits for `hupReady` to get closed anyway.

Comment (grobie) [reply cmd/prometheus/main.go]: Ah, I had missed that line. :+1:

File: storage/tsdb/tsdb.go [hunk 6]
```diff
@@ -27,6 +28,63 @@ import (
...
+func (s *ReadyStorage) Get() *tsdb.DB {
+	if x := s.get(); x != nil {
+		return x.db
+	}
+	return nil
+}
+
+func (s *ReadyStorage) get() *adapter {
+	s.mtx.RLock()
+	x := s.a
```
Comment (grobie) [inline storage/tsdb/tsdb.go]: Interesting. What's the reason to have this additional variable assignment? I'd have naively written:

```go
s.mtx.RLock()
defer s.mtx.RUnlock()
return s.a
```

Comment (fabxc) [reply storage/tsdb/tsdb.go]: Yea, that's probably fine. I've just had a few issues in the last couple of months where `defer` had a very significant performance penalty. Just got in a habit of skipping on it if there was no legitimate concern of intermediate panics.

Comment (grobie) [reply storage/tsdb/tsdb.go]: Thanks for the info! Was just curious.

**Pull request merged**